### ISSUE TITLE
Usage of Readonly types instead of mutable in TS typings

### ISSUE THREAD

**Issue by PinkaminaDianePie** [prisma/prisma-client-js#497]
Currently all arrays in input positions use `Array` type and it leads to a lot of issues, such as incompatibility with Facebook `DataLoader` library. DataLoader expect keys to be a ReadonlyArray, while Prisma expects `{where: { id: { in: keys } }}` to accept mutable Array. I think it's safe to change types of arrays in input positions to `ReadonlyArray` as well (i believe Prisma doesn't mutate them under the hood, so it's safe to do that), while keeping types in output positions as `Array` (we shouldn't enforce users to keep result array immutable, so we can't use `ReadonlyArray` here).
If I understand correctly, currently we use only one generic for all input types: `export declare type Enumerable<T> = T | Array<T>;`, so changing it to `export declare type Enumerable<T> = T | ReadonlyArray<T>;
` should be enough.

It would be nice to have immutable types for objects as well, like 
```typescript
export declare type UserCreateInput = Readonly<{
    id?: string | null;
    name: string;
}>;
```
instead of 
```typescript
export declare type UserCreateInput = {
    id?: string | null;
    name: string;
};
```
because of the same reasons as described above. If Prisma doesn't mutate this object, it's better to mark it as immutable, so user will be able to pass both mutable/immutable objects, not only mutable ones like right now

**Comment by reiv** [comment]
Seeing as it's been several months now - is this still being considered? At least the addition of `ReadonlyArray<T>` to the definition of `Enumerable` as @PinkaminaDianePie suggested seems like a trivial and non-breaking change.

I suppose in the meantime one could cook up a quick and dirty regex to post-process the generated index.d.ts, but that's obviously not optimal.

**Comment by Sytten** [comment]
Yeah I faced the same issue the other day, minor annoyance that could be easily fixed @timsuchanek

**Comment by timsuchanek** [comment]
Thanks everyone for your comments!
After getting seduced to a quick fix, we realized, that this has negative consequences for other use-cases that we don't want to introduce.

Example:

```ts
import { IntFilter, PrismaClient } from '@prisma/client'

async function main() {
  const prisma = new PrismaClient()

  const ids: number[] = []

  const filter: IntFilter = {
    in: []
  }

  if (ids && Array.isArray(filter.in)) {
    filter.in.push(...ids)
  }

  const data = await prisma.user.findMany({
    where: {
      id: filter
    }
  })

  console.log(data)
  prisma.$disconnect()
}
```

In words: It takes the opportunity to programmatically, type-safe construct arguments.
The way to go here is to already assign the Prisma input type like `IntFilter` to the variable and then mutate the object as needed. That's a very common pattern (not just with arrays) that's being used with Prisma Client, which we still want to allow.

**Comment by Sytten** [comment]
Agreed but what about setting Enumerate to
```typescript
export type Enumerable<T> = T | Array<T> | ReadonlyArray<T>
```
That would solve my issue with dataloaders. Dataloaders use a ReadonlyArray in the load method so currently I have to cast it to an array. https://github.com/graphql/dataloader/blob/master/src/index.d.ts#L74
